bitar/src/api/compress.rs

```diff
@@ -0,0 +1,265 @@
+use std::error;
+use std::fmt;
+use std::path::PathBuf;
+
+use blake2::{Blake2b512, Digest};
+use futures_util::{future, StreamExt};
+use tokio::fs;
+use tokio::io;
+use tokio::io::AsyncSeekExt;
+use tokio::io::{AsyncRead, AsyncWrite, AsyncWriteExt};
+use tokio::task::JoinError;
+
+use crate::chunk_dictionary;
+use crate::chunker;
+use crate::Compression;
+use crate::CompressionAlgorithm;
+
+pub const PKG_VERSION: &str = env!("CARGO_PKG_VERSION");
+
+/// Options for the `create_archive` function
+#[derive(Clone, Debug)]
+pub struct CreateArchiveOptions {
+    /// The configuration to use when creating a chunk stream from the input
+    pub chunker_config: chunker::Config,
+
+    /// Number of parallel buffers to use when manipulating chunks
+    pub num_chunk_buffers: usize,
+
+    /// The length that the chunk hash should be truncated to for the output
+    pub chunk_hash_length: usize,
+
+    /// A temporary file is used to write intermediate chunk data. Setting this
+    /// option forces this file to be used instead of a randomly generated one
+    pub temporary_file_override: Option<PathBuf>,
+
+    /// The type of compression to use when compressing a chunk
+    pub compression: Option<Compression>,
+}
+
+impl Default for CreateArchiveOptions {
+    fn default() -> CreateArchiveOptions {
+        let num_buffers = match num_cpus::get() {
+            0 | 1 => 1,
+            n => n * 2,
+        };
+
+        CreateArchiveOptions {
+            chunker_config: chunker::Config::RollSum(chunker::FilterConfig::default()),
+            num_chunk_buffers: num_buffers,
+            chunk_hash_length: 64,
+            temporary_file_override: None,
+            compression: Some(Compression {
+                algorithm: CompressionAlgorithm::Brotli,
+                level: 6,
+            }),
+        }
+    }
+}
+
+/// Output from the `create_archive` function
+#[derive(Default)]
+pub struct CreateArchiveResult {
+    pub source_hash: Vec<u8>,
+    pub source_length: usize,
+    pub header: chunk_dictionary::ChunkDictionary,
+}
+
+/// Error from the `create_archive` function
+#[derive(Debug)]
+pub enum CreateArchiveError {
+    /// Error that occurred while operatin on the temp file used to creat the archive
+    TempFileError(io::Error),
+    /// Failed to chunk the input file
+    ChunkerError(JoinError),
+    /// Failed to write to the output file
+    OutputWriteError(io::Error),
+}
+
+impl fmt::Display for CreateArchiveError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            CreateArchiveError::TempFileError(_) => {
+                write!(f, "Error occurred while operating on the temp file")
+            }
+            CreateArchiveError::ChunkerError(_) => write!(f, "Error chunking the input file"),
+            CreateArchiveError::OutputWriteError(_) => write!(f, "Error writing to output file"),
+        }
+    }
+}
+
+impl error::Error for CreateArchiveError {}
```

Comment:
```suggestion
impl error::Error for CreateArchiveError {}
```
```suggestion
impl error::Error for CreateArchiveError {
    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
        ...
    }
}
```
Might want to implement the source() method of Error too for error traversing.

Reply:
Makes sense, added.